### PR TITLE
Expandable Bottom Sheet 2 Electric Boogaloo

### PR DESCRIPTION
**Pull request by caseycrogers**
## :scroll: Description

Replaces #89 as the latter was so out of date it was easier to start over than to merge.

I'm going to self-review this a bit, but I think it's ready to review and merge? It's dependent on #147 so that'll have to go in first.

## :pencil: Checklist

- [ ] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated the docs if needed
- [ ] All tests passing
- [ ] No breaking changes

### CODE REVIEW

File: feedback/lib/src/feedback_widget.dart [hunk 1]
```diff
@@ -117,117 +132,153 @@ class FeedbackWidgetState extends State<FeedbackWidget>
...
-                            BetterFeedback.of(context).onFeedback!,
-                            screenshotController,
-                            feedback,
-                            widget.pixelRatio,
-                            extras: extras,
-                          );
-                          painterController.clear();
-                        },
+                builder: (context, screenshotChild) {
+                  return CustomMultiChildLayout(
```
Comment (caseycrogers) [inline feedback/lib/src/feedback_widget.dart:167]: This section has gotten huge-it'd probably be a good candidate for breaking out into a separate widget. I could add that to this PR or we could put it off to a future PR.

File: feedback/lib/src/feedback_widget.dart [hunk 2]
```diff
@@ -117,117 +132,153 @@ class FeedbackWidgetState extends State<FeedbackWidget>
...
+                    children: [
+                      LayoutId(
+                        id: _controlsColumnId,
+                        child: Padding(
+                          padding: EdgeInsets.only(left: padding),
+                          child: ScaleAndFade(
```
Comment (caseycrogers) [inline feedback/lib/src/feedback_widget.dart:173]: This guy just adds a nice fade away effect on the background as the sheet is dragged up.

File: feedback/lib/src/better_feedback.dart [hunk 3]
```diff
@@ -18,7 +18,49 @@ typedef OnSubmit = void Function(
...
+/// can be dragged to expand it.
+///
+/// It should be placed in a stack over the sheet's scrollable element so that
+/// users can click and drag on it-the drag handle ignores pointers so the drag
+/// will pass through to the scrollable beneath.
+///
+/// TODO(caseycrogers): Replace this with a pre-built drag handle above the
+///   builder function once `DraggableScrollableController` is available in
+///   production.
+///   See: https://github.com/flutter/flutter/pull/92440.
```
Comment (ueman) [inline feedback/lib/src/better_feedback.dart:47]: This shouldn't be a part of the doc comments.
```suggestion
// TODO(caseycrogers): Replace this with a pre-built drag handle above the
//   builder function once `DraggableScrollableController` is available in
//   production.
//   See: https://github.com/flutter/flutter/pull/92440.
```
Maybe we can already create an issue for that, so we don't forget.

File: feedback/lib/src/better_feedback.dart [hunk 4]
```diff
@@ -53,7 +95,7 @@ class BetterFeedback extends StatefulWidget {
   const BetterFeedback({
     Key? key,
     required this.child,
-    this.feedbackBuilder,
+    this.feedbackSheetBuilder,
```
Comment (ueman) [inline feedback/lib/src/better_feedback.dart:98]: Can we try to not do a breaking change and instead just deprecate the old callback but keep it working?
Overall, this is just a visual / behavioral change, right?

File: feedback/lib/src/feedback_bottom_sheet.dart [hunk 5]
```diff
@@ -2,34 +2,140 @@
...
-    // `BetterFeedback` is used above `MaterialApp` in the widget tree so that
-    // the nested navigation in navigate mode works properly.
-    return Navigator(
-      onGenerateRoute: (_) => MaterialPageRoute<void>(
-        builder: (context) => Material(
+    // We need to supply an navigator so that the contents of the bottom sheet
+    // have access to it. Overlays are used by many material widgets
+    // such as `TextField` and `DropDownButton`. Typically, this would be
+    // provided by `MaterialApp`, but `BetterFeedback` is above `MaterialApp` in
+    // the widget tree.
```
Comment (ueman) [inline feedback/lib/src/feedback_bottom_sheet.dart:27]: This comment seems now out of date, right? Don't we need a Navigator anymore?

File: feedback/lib/src/feedback_widget.dart [hunk 6]
```diff
@@ -46,8 +51,18 @@ class FeedbackWidget extends StatefulWidget {
...
+  // Padding to put around the interactive screenshot preview.
+  final double padding = 8;
+
   final BackButtonInterceptor _interceptor = BackButtonInterceptor();
 
+  /// We use a ValueNotifier instead of just a double and `SetState` because
+  /// rebuilding the feedback sheet mid-drag cancels the drag.
+  /// TODO(caseycrogers): replace `sheetProgress` with a direct reference to
+  ///   `DraggableScrollableController` when the latter gets into production.
+  ///   See: https://github.com/flutter/flutter/pull/92440.
```
Comment (ueman) [inline feedback/lib/src/feedback_widget.dart:63]:
```suggestion
  // TODO(caseycrogers): replace `sheetProgress` with a direct reference to
  // `DraggableScrollableController` when the latter gets into production.
  // See: https://github.com/flutter/flutter/pull/92440.
```

**Comment by ueman** [comment]
I'll hopefully have some time over the next few days and test this more throughout. Thanks again!

File: feedback/lib/src/better_feedback.dart (hunk 3 above)
Comment (caseycrogers) [reply feedback/lib/src/better_feedback.dart]: https://github.com/ueman/feedback/issues/150

File: feedback/lib/src/better_feedback.dart (hunk 4 above)
Comment (caseycrogers) [reply feedback/lib/src/better_feedback.dart]: Yeah this was just visual, as I was working I kept getting confused as to what `feedbackBuilder` was. I'll snip this change from this PR-it belongs in it's own PR anyway (if we decide to do it).

File: feedback/lib/src/feedback_bottom_sheet.dart (hunk 5 above)
Comment (caseycrogers) [reply feedback/lib/src/feedback_bottom_sheet.dart]: FIxed.

File: feedback/lib/src/feedback_widget.dart (hunk 6 above)
Comment (caseycrogers) [reply feedback/lib/src/feedback_widget.dart]: #150

**Review by caseycrogers — COMMENTED** [review]
Fixed all the comments!

File: feedback/lib/src/feedback_widget.dart (hunk 1 above)
Comment (ueman) [reply feedback/lib/src/feedback_widget.dart:167]: Your call

File: feedback/lib/src/feedback_widget.dart (hunk 2 above)
Comment (ueman) [reply feedback/lib/src/feedback_widget.dart:173]: It looks really good 👍

**Comment by ueman** [comment]
Dragging seems to be broken on desktop. I haven't figured out why, though. On mobile it's really awesome.

**Comment by caseycrogers** [comment]
Hmm, let me see if I can figure it out on my end. On Windows?

**Comment by ueman** [comment]
> Hmm, let me see if I can figure it out on my end. On Windows?

I've tried on a Mac, but Windows is probably also affected. Maybe also web?

Also, I've just discovered another error: If the bottom sheet covers the whole screen and if you're then scrolling the text input to the top out of screen, you'll get a `Stack Overflow` exception.

**Comment by caseycrogers** [comment]
> > Hmm, let me see if I can figure it out on my end. On Windows?
> 
> I've tried on a Mac, but Windows is probably also affected. Maybe also web?
> 
> Also, I've just discovered another error: If the bottom sheet covers the whole screen and if you're then scrolling the text input to the top out of screen, you'll get a `Stack Overflow` exception.
> 
> Also the bottom sheet needs to be closed before taking the screenshot, otherwise it doesn't work.

As for desktop, I tried on Windows. The error should also exist for web. What's happening is `DraggableScrollableSheet` only works for dragging (not scrolling) and you can only scroll on those platforms.
I could put a PR against Flutter to fix this, but I think it's a little more complicated:

depending on the user's scroll directions (depends on platform and settings) and whether or not they're using a mouse with a scroll wheel or two finger drag on a touchpad, the sheet expanding will feel really jarring and unintuitive. 

Once the `DraggableScrollableController` PR makes it into production Flutter (I think in about a month?) we could make the drag handle draggable on those platforms so even if scrolling won't expand the sheet, clicking and dragging on the handle would.

In the meantime, I propose the following: only allow the draggable sheet on mobile. We add a check somewhere where if it detects that the platform is not mobile and `sheetIsDraggable` is true, it'll print a warning to console and treat `sheeIsDraggable` as false. We can make the current default behavior platform dependent.

I'll look into the stack overflow and the screenshot issue.

**Comment by caseycrogers** [comment]
All the errors should be resolved. Sorry this was initially so buggy, I was lot more thorough this round.

The golden images are pretty badly garbled, I'm not sure what's going on there. Do you think you could take a crack at it?

File: feedback_gitlab/example/example.dart [hunk 7]
```diff
@@ -2,7 +2,7 @@ import 'package:feedback_gitlab/feedback_gitlab.dart';
 import 'package:flutter/material.dart';
 
 void main() {
-  runApp(const BetterFeedback(child: MyApp()));
```
Comment (ueman) [inline feedback_gitlab/example/example.dart:5]: This should be reverted

**Comment by ueman** [comment]
See #154 for what was wrong with the golden images

**Pull request merged**